pkg/pool-manager/virtualmachine_pool.go

```diff
@@ -445,51 +540,123 @@ func (p *PoolManager) vmWaitingCleanupLook() {
 	for _ = range c {
 		p.poolMutex.Lock()
 
-		configMap, err := p.kubeClient.CoreV1().ConfigMaps(p.managerNamespace).Get(context.TODO(), names.WAITING_VMS_CONFIGMAP, metav1.GetOptions{})
-		if err != nil {
-			logger.Error(err, "failed to get config map", "configMapName", names.WAITING_VMS_CONFIGMAP)
-			p.poolMutex.Unlock()
+		p.handleStaleLegacyConfigMapEntries(logger)
+
+		p.handleStaleMacEntries(logger)
+
+		p.poolMutex.Unlock()
+	}
+}
+
+func (p *PoolManager) handleStaleMacEntries(parentLogger logr.Logger) error {
+	logger := parentLogger.WithName("handleStaleMacEntries")
+	var macMapChanged bool
+	for macAddress, macEntry := range p.macPoolMap {
+		macMapChanged = false
+		if !macEntry.hasPendingTransaction() {
 			continue
 		}
+		isEntryStale, err := p.isMacTransactionStale(macAddress)
+		if err == nil && isEntryStale {
+			macMapChanged = true
+			logger.Info("entry is stale", "macAddress", macAddress, "vmFullName", macEntry.instanceName, "interfaceName", macEntry.macInstanceKey, "stale TS", macEntry.transactionTimestamp)
+			vm, err := p.getvmInstance(macEntry.instanceName)
+			if err != nil && apierrors.IsNotFound(err) {
+				logger.Info("vm no longer exists or does not have Template. removing mac from pool", "macAddress", macAddress, "entry", macEntry)
+				p.macPoolMap.removeMacEntry(macAddress)
+			} else if err == nil {
+				p.macPoolMap.alignMacEntryAccordingToVmInterface(macAddress, getVirtualMachineInterfaces(vm))
+			} else {
+				return err
+			}
```

Comment:
Let's reduce the amount of nested code and simplify conditions
```golang
if err != nil {
    if apierrors.IsNotFound(err) {
       .....
       return nil
    } else {
      return err
    }
}
p.macPoolMap.alignMacEntryAccordingToVmInterface(macAddress, getVirtualMachineInterfaces(vm))
```

Reply:
I can't return `nil` in case of not found, I can `continue` instead. is that better?
```
			vm, err := p.getvmInstance(macEntry.instanceName)
			if err != nil {
				if apierrors.IsNotFound(err) {
					logger.V(1).Info("vm no longer exists or does not have Template. removing mac from pool", "macAddress", macAddress, "entry", macEntry)
					macsToRemove = append(macsToRemove, macAddress)
					continue
				} else {
					return err
				}
			}
			macsToAlign[macAddress] = *vm
```